tsl/src/compression/compression.c

```diff
@@ -1550,20 +1552,36 @@ row_decompressor_decompress_row_to_table(RowDecompressor *decompressor)
 	 * insert the entire batch into one index, then into another, and so on.
 	 * Working with one index at a time gives better data access locality,
 	 * which reduces the load on shared buffers cache.
+	 *
 	 * The normal Postgres code inserts each row into all indexes, so to do it
 	 * the other way around, we create a temporary ResultRelInfo that only
-	 * references one index. Then we loop over indexes, and for each index we
-	 * set it to this temporary ResultRelInfo, and insert all rows into this
-	 * single index.
+	 * references one index. We need to re-open the index for the relation
+	 * here since the size of ResultRelInfo can change (this happened in 17.1)
+	 * and a blind copy of the ResultRelInfo will not copy the correct data.
+	 *
+	 * Then we loop over indexes, and for each index we set it to this
+	 * temporary ResultRelInfo, and insert all rows into this single index.
 	 */
 	if (decompressor->indexstate->ri_NumIndices > 0)
 	{
-		ResultRelInfo indexstate_copy = *decompressor->indexstate;
 		Relation single_index_relation;
 		IndexInfo *single_index_info;
-		indexstate_copy.ri_NumIndices = 1;
-		indexstate_copy.ri_IndexRelationDescs = &single_index_relation;
-		indexstate_copy.ri_IndexRelationInfo = &single_index_info;
+
+		/* We overallocate memory for ResultRelInfo copy here because 17.1
+		 * added an extra field and when calling InitResultRelInfo() on a
+		 * ResultRelInfo object it will attempt to write after the end of the
+		 * structure allocated in the 17.0 server. */
+		union OverallocatedResultRelInfo
```

Comment:
Why do we even need this to be stack allocated? Isn't it better to just heap-allocate it as that would be safer given that palloc already over-allocates? Or, if we are not satisfied with that, we can just give palloc a bigger size.